File: tpu_sync/weight_sync/weight_synchronization_worker_service_client.cc
// Copyright 2026 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "tpu_sync/weight_sync/weight_synchronization_worker_service_client.h"

#include <memory>
#include <string>
#include <utility>
#include <vector>

#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/string_view.h"
#include "absl/time/clock.h"
#include "absl/time/time.h"
#include "grpcpp/client_context.h"
#include "grpcpp/create_channel.h"
#include "grpcpp/security/credentials.h"
#include "grpcpp/support/status.h"
#include "xla/tsl/concurrency/future.h"
#include "tpu_sync/rpc/raiden_service.grpc.pb.h"
#include "tpu_sync/rpc/raiden_service.pb.h"

namespace tpu_raiden {
namespace weight_sync {
namespace {

void SetContextDeadline(grpc::ClientContext* context, absl::Duration timeout) {
  if (timeout < absl::InfiniteDuration() && timeout > absl::ZeroDuration()) {
    context->set_deadline(absl::ToChronoTime(absl::Now() + timeout));
  }
}

}  // namespace

WeightSynchronizationWorkerServiceClient::
    WeightSynchronizationWorkerServiceClient(
        std::shared_ptr<::grpc::ChannelInterface> channel)
    : stub_(::tpu_sync::rpc::WeightSynchronizationWorkerService::NewStub(
          channel)) {}

WeightSynchronizationWorkerServiceClient::
    WeightSynchronizationWorkerServiceClient(
        std::unique_ptr<
            ::tpu_sync::rpc::WeightSynchronizationWorkerService::StubInterface>
            stub)
    : stub_(std::move(stub)) {}

WeightSynchronizationWorkerServiceClient::
    WeightSynchronizationWorkerServiceClient(absl::string_view target)
    : stub_(::tpu_sync::rpc::WeightSynchronizationWorkerService::NewStub(
          ::grpc::CreateChannel(std::string(target),
                                ::grpc::InsecureChannelCredentials()))) {}

tsl::Future<::tpu_sync::rpc::ControlResponse>
WeightSynchronizationWorkerServiceClient::HandleControl(
    const ::tpu_sync::rpc::ControlRequest& request, absl::Duration timeout) {
  auto [promise, future] = tsl::MakePromise<::tpu_sync::rpc::ControlResponse>();
  auto context = std::make_shared<::grpc::ClientContext>();
  SetContextDeadline(context.get(), timeout);

  auto response = std::make_shared<::tpu_sync::rpc::ControlResponse>();

  stub_->async()->HandleControl(
      context.get(), &request, response.get(),
      [context, response,
       promise = std::move(promise).ToShared()](::grpc::Status status) mutable {
        if (!status.ok()) {
          promise->Set(
              absl::Status(static_cast<absl::StatusCode>(status.error_code()),
                           absl::StrCat("HandleControl RPC failed: ",
                                        status.error_message())));
        } else {
          promise->Set(std::move(*response));
        }
      });

  return future;
}

tsl::Future<::tpu_sync::rpc::ControlResponse>
WeightSynchronizationWorkerServiceClient::StartTransfer(
    const ::tpu_sync::rpc::StartTransferRequest& start_transfer_request,
    const std::vector<std::string>& peers, absl::Duration timeout) {
  ::tpu_sync::rpc::ControlRequest req;
  req.set_command(::tpu_sync::rpc::ControlRequest::COMMAND_START_TRANSFER);
  *req.mutable_start_transfer_request() = start_transfer_request;
  for (const auto& peer : peers) {
    req.add_peers(peer);
  }

  return HandleControl(req, timeout);
}

tsl::Future<::tpu_sync::rpc::ControlResponse>
WeightSynchronizationWorkerServiceClient::Shutdown(absl::Duration timeout) {
  ::tpu_sync::rpc::ControlRequest req;
  req.set_command(::tpu_sync::rpc::ControlRequest::COMMAND_SHUTDOWN);

  return HandleControl(req, timeout);
}

}  // namespace weight_sync
}  // namespace tpu_raiden

File: tpu_sync/weight_sync/weight_synchronization_worker_service_client.py
# Copyright 2026 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Python client for WeightSynchronizationWorkerService."""

from typing import Optional, Sequence
import grpc
from tpu_sync.rpc import raiden_service_pb2
from tpu_sync.rpc import raiden_service_pb2_grpc


class WeightSynchronizationWorkerServiceClient:
  """Client for interacting with WeightSynchronizationWorkerService."""

  def __init__(
      self,
      target: Optional[str] = None,
      *,
      channel: Optional[grpc.Channel] = None,
      stub: Optional[
          raiden_service_pb2_grpc.WeightSynchronizationWorkerServiceStub
      ] = None,
      options: Optional[Sequence[tuple[str, object]]] = None,
  ):
    """Initializes the client.

    Args:
      target: Server address string (e.g. 'localhost:50051' or 'ip:port').
      channel: Optional existing grpc.Channel to use.
      stub: Optional existing WeightSynchronizationWorkerServiceStub to use.
      options: Optional list of key-value pairs to configure the grpc channel if
        creating a new channel.
    """
    if stub is not None:
      self._stub = stub
      self._channel = channel
      self._owns_channel = False
    elif channel is not None:
      self._channel = channel
      self._stub = (
          raiden_service_pb2_grpc.WeightSynchronizationWorkerServiceStub(
              channel
          )
      )
      self._owns_channel = False
    elif target is not None:
      self._channel = grpc.insecure_channel(target, options=options)
      self._stub = (
          raiden_service_pb2_grpc.WeightSynchronizationWorkerServiceStub(
              self._channel
          )
      )
      self._owns_channel = True
    else:
      raise ValueError("One of target, channel, or stub must be provided.")

  @property
  def stub(
      self,
  ) -> raiden_service_pb2_grpc.WeightSynchronizationWorkerServiceStub:
    """Returns the underlying gRPC stub."""
    return self._stub

  @property
  def channel(self) -> Optional[grpc.Channel]:
    """Returns the underlying gRPC channel if available."""
    return self._channel

  def handle_control(
      self,
      request: raiden_service_pb2.ControlRequest,
      timeout: Optional[float] = None,
  ) -> grpc.Future:
    """Executes an asynchronous HandleControl RPC.

    Args:
      request: The ControlRequest proto to send.
      timeout: Optional timeout in seconds.

    Returns:
      grpc.Future representing the pending ControlResponse.
    """
    return self._stub.HandleControl.future(request, timeout=timeout)

  def start_transfer(
      self,
      start_transfer_request: raiden_service_pb2.StartTransferRequest,
      peers: Optional[Sequence[str]] = None,
      timeout: Optional[float] = None,
  ) -> grpc.Future:
    """Convenience helper to initiate weight transfer asynchronously.

    Args:
      start_transfer_request: StartTransferRequest proto specifying transfer
        parameters and schedules.
      peers: Optional list of destination peer endpoints.
      timeout: Optional timeout in seconds.

    Returns:
      grpc.Future representing the pending ControlResponse.
    """
    req = raiden_service_pb2.ControlRequest(
        command=raiden_service_pb2.ControlRequest.COMMAND_START_TRANSFER,
        start_transfer_request=start_transfer_request,
        peers=peers or [],
    )
    return self.handle_control(req, timeout=timeout)

  def shutdown(self, timeout: Optional[float] = None) -> grpc.Future:
    """Convenience helper to instruct the remote worker to shut down asynchronously.

    Args:
      timeout: Optional timeout in seconds.

    Returns:
      grpc.Future representing the pending ControlResponse.
    """
    req = raiden_service_pb2.ControlRequest(
        command=raiden_service_pb2.ControlRequest.COMMAND_SHUTDOWN
    )
    return self.handle_control(req, timeout=timeout)

  def close(self) -> None:
    """Closes the underlying gRPC channel if owned by this client."""
    if self._owns_channel and self._channel is not None:
      self._channel.close()

  def __enter__(self) -> "WeightSynchronizationWorkerServiceClient":
    return self

  def __exit__(self, exc_type, exc_val, exc_tb) -> None:
    self.close()
